ethcore/src/blockchain/blockchain.rs

```diff
@@ -375,50 +375,56 @@ impl BlockChain {
 		while current_from != current_to {
 			from_branch.push(current_from);
 			current_from = from_details.parent.clone();
-			from_details = self.block_details(&from_details.parent).expect(&format!("4. Expected to find details for block {:?}", from_details.parent));
+			from_details = try!(self.block_details(&from_details.parent).ok_or(FatalError::MissingBlockDetails { hash: from_details.parent.clone() }));
 
 			to_branch.push(current_to);
 			current_to = to_details.parent.clone();
-			to_details = self.block_details(&to_details.parent).expect(&format!("5. Expected to find details for block {:?}", from_details.parent));
+			to_details = try!(self.block_details(&to_details.parent).ok_or(FatalError::MissingBlockDetails { hash: to_details.parent.clone() }));
 		}
 
 		let index = from_branch.len();
 
 		from_branch.extend(to_branch.into_iter().rev());
 
-		TreeRoute {
+		Ok(TreeRoute {
 			blocks: from_branch,
 			ancestor: current_from,
 			index: index
-		}
+		})
 	}
 
 	/// Inserts the block into backing cache database.
 	/// Expects the block to be valid and already verified.
 	/// If the block is already known, does nothing.
-	pub fn insert_block(&self, bytes: &[u8], receipts: Vec<Receipt>) {
+	/// Returns Option<H256> - newly inserted block hash if insert was successful
+	pub fn insert_block(&self, bytes: &[u8], receipts: Vec<Receipt>) -> Result<Option<H256>, Error> {
```

Comment:
Is it guaranteed that the blockchain remains in consistent state after returning an error from this function?